perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java

```diff
@@ -2060,4 +2179,108 @@ private void closeExtSourcesAfterSynchronization(ExtSource membersSource, ExtSou
 			}
 		}
 	}
+	
+	@Override
+	public void processRelationMembers(PerunSession sess, Group resultGroup, List<Member> changedMembers, int sourceGroupId, boolean addition) throws GroupOperationsException {
+
+		List<Member> newMembers;
+
+		try {
+			if (addition) {
+				newMembers = addIndirectMembers(sess, resultGroup, changedMembers, sourceGroupId);
+			} else {
+				newMembers = removeIndirectMembers(sess, resultGroup, changedMembers, sourceGroupId);
+			}
+
+			if (newMembers.isEmpty()) {
+				return;
+			}
+
+			List<Integer> relations = groupsManagerImpl.getGroupRelations(sess, resultGroup.getId());
+			for (Integer groupId : relations) {
+				processRelationMembers(sess, groupsManagerImpl.getGroupById(sess, groupId), newMembers, resultGroup.getId(), addition);
+			}
+		} catch (WrongReferenceAttributeValueException | WrongAttributeValueException | AlreadyMemberException |
+				InternalErrorException | NotGroupMemberException | GroupNotExistsException ex) {
+			throw new GroupOperationsException(ex);
+		}
+	}
+
+	@Override
+	public Group createGroupUnion(PerunSession sess, Group resultGroup, Group operandGroup, boolean parentFlag) throws GroupOperationsException, InternalErrorException {
+		
+		//check if any of the groups is members group
+		if(resultGroup.getName().equals(VosManager.MEMBERS_GROUP) || operandGroup.getName().equals(VosManager.MEMBERS_GROUP)) {
+			throw new InternalErrorException("Union cannot be created on members group.");
+		}
+
+		// check if both groups are from same VO
+		if (resultGroup.getVoId() != operandGroup.getVoId()) {
+			throw new InternalErrorException("Groups are not from same VO");
+		}
+
+		// check if result group is the same as operand group
+		if (resultGroup.getId() == operandGroup.getId()) {
+			throw new InternalErrorException("Result group cannot be the same as operand group.");
+		} 
+
+		// check if there is already a record of these two groups
+		if (this.groupsManagerImpl.isRelationBetweenGroups(resultGroup, operandGroup)) {
+			throw new InternalErrorException("There is already an operation defined between these two groups: " + resultGroup + " and " + operandGroup 
+					+ " or they are in hierarchical structure.");
+		}
+
+		// check cycle between groups
+		if (checkGroupsCycle(sess, resultGroup.getId(), operandGroup.getId())) {
+			throw new InternalErrorException("This relation would create group transitivity.");
```

Comment:
Unless there is another check which prevent user this condition, another exception should be used.